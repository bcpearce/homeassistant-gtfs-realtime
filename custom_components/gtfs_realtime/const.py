"""Constants for the GTFS Realtime integration."""

DOMAIN = "gtfs_realtime"

CONF_GTFS_PROVIDER = "gtfs_provider"
CONF_API_KEY = "api_key"
CONF_GTFS_STATIC_DATA = "gtfs_static_data"
CONF_URL_ENDPOINTS = "url_endpoints"
CONF_ROUTE_ICONS = "route_icons"
CONF_ROUTE_IDS = "route_ids"
CONF_STOP_IDS = "stop_ids"
CONF_ARRIVAL_LIMIT = "arrival_limit"
STOP_ID = "stop_id"
ROUTE_ID = "route_id"

SSI_DB = "station_stop_info_db"
TI_DB = "trip_info_db"
CAL_DB = "calendar_db"
RTI_DB = "route_info_db"

COORDINATOR_REALTIME = "coordinator_realtime"
COORDINATOR_STATIC = "coordinator_static"

PLATFORM = "platform"

"""These Constants are user facing."""

TRIP_ID_PRETTY = "Trip ID"
HEADSIGN_PRETTY = "Headsign"
ROUTE_COLOR_PRETTY = "Route Color"
ROUTE_TEXT_COLOR_PRETTY = "Route Text Color"
ROUTE_TYPE_PRETTY = "Route Type"

HEADER_PRETTY = "Header"
DESCRIPTION_PRETTY = "Description"
