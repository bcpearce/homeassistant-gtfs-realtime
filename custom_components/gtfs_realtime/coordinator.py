"""GTFS Realtime Coordinator."""

import asyncio
from collections.abc import Iterable
from datetime import datetime, timedelta
import logging
import os

from gtfs_station_stop.calendar import Calendar
from gtfs_station_stop.feed_subject import FeedSubject
from gtfs_station_stop.route_info import RouteInfoDatabase
from gtfs_station_stop.static_database import async_factory
from gtfs_station_stop.station_stop_info import StationStopInfoDatabase
from gtfs_station_stop.trip_info import TripInfoDatabase
from homeassistant.core import HomeAssistant
from homeassistant.helpers.update_coordinator import DataUpdateCoordinator

_LOGGER = logging.getLogger(__name__)


class GtfsRealtimeCoordinator(DataUpdateCoordinator):
    """GTFS Realtime Update Coordinator."""

    def __init__(
        self,
        hass: HomeAssistant,
        feed_subject: FeedSubject,
        gtfs_static_zip: Iterable[os.PathLike] | os.PathLike | None = None,
        **kwargs,
    ) -> None:
        """Initialize the GTFS Update Coordinator to notify all entities upon poll."""
        self.static_timedelta: timedelta = kwargs.get(
            "static_timedelta", timedelta(hours=24)
        )
        self.realtime_timedelta: timedelta = kwargs.get(
            "realtime_timedelta", timedelta(seconds=60)
        )
        super().__init__(
            hass,
            _LOGGER,
            name="GTFS Realtime",
            update_interval=self.realtime_timedelta,
        )
        self.hub = feed_subject
        self.gtfs_static_zip = gtfs_static_zip
        self.calendar: Calendar | None = None
        self.station_stop_info_db: StationStopInfoDatabase | None = None
        self.trip_info_db: TripInfoDatabase | None = None
        self.route_info_db: RouteInfoDatabase | None = None
        self.kwargs = kwargs
        self.services_len: int = 0
        self.stop_infos_len: int = 0
        self.trip_infos_len: int = 0
        self.route_infos_len: int = 0
        self.last_static_update: datetime | None = None
        _LOGGER.info("Setup GTFS Realtime Update Coordinator")
        _LOGGER.info(f"Realtime GTFS update interval {self.realtime_timedelta}")
        _LOGGER.info(f"Static GTFS update interval {self.static_timedelta}")

    async def _async_update_data(self):
        """Fetch data from API endpoint."""
        await self.hub.async_update()
        # Update the static resource if it is past that timedelta
        if (
            self.last_static_update is None
            or datetime.now() - self.last_static_update > self.static_timedelta
        ):
            await self.async_update_static_data()

    async def async_update_static_data(self):
        async with asyncio.TaskGroup() as tg:
            cal_db_task = tg.create_task(
                async_factory(Calendar, *self.gtfs_static_zip, **self.kwargs)
            )
            ssi_db_task = tg.create_task(
                async_factory(
                    StationStopInfoDatabase, *self.gtfs_static_zip, **self.kwargs
                )
            )
            ti_db_task = tg.create_task(
                async_factory(TripInfoDatabase, *self.gtfs_static_zip, **self.kwargs)
            )
            rti_db_task = tg.create_task(
                async_factory(RouteInfoDatabase, *self.gtfs_static_zip, **self.kwargs)
            )
        if any(
            db is None
            for db in (
                self.calendar,
                self.station_stop_info_db,
                self.trip_info_db,
                self.route_info_db,
            )
        ):
            (
                self.calendar,
                self.station_stop_info_db,
                self.trip_info_db,
                self.route_info_db,
            ) = (
                cal_db_task.result(),
                ssi_db_task.result(),
                ti_db_task.result(),
                rti_db_task.result(),
            )
            self.services_len = len(self.calendar.services)
            self.stop_infos_len = len(self.station_stop_info_db.station_stop_infos)
            self.trip_infos_len = len(self.trip_info_db.trip_infos)
            self.route_infos_len = len(self.route_info_db.route_infos)
            _LOGGER.info(
                f"GTFS Static Coordinator Initial Update Complete {self.gtfs_static_zip}"
            )
            _LOGGER.info(f"GTFS Static Coordinator Services: {self.services_len}")
            _LOGGER.info(
                f"GTFS Static Coordinator Station Stop Infos: {self.stop_infos_len}"
            )
            _LOGGER.info(f"GTFS Static Coordinator Trip Infos: {self.trip_infos_len}")
            _LOGGER.info(f"GTFS Static Coordinator Route Infos: {self.route_infos_len}")
        else:
            old_services_len = self.services_len
            self.calendar.services = (
                self.calendar.services | cal_db_task.result().services
            )
            self.services_len = len(self.calendar.services)

            old_stop_infos_len = self.stop_infos_len
            self.station_stop_info_db.station_stop_infos = (
                self.station_stop_info_db.station_stop_infos
                | (ssi_db_task.result().station_stop_infos)
            )
            self.stop_infos_len = len(self.station_stop_info_db.station_stop_infos)

            old_trip_infos_len = self.trip_infos_len
            self.trip_info_db.trip_infos = (
                self.trip_info_db.trip_infos | ti_db_task.result().trip_infos
            )
            self.trip_infos_len = len(self.trip_info_db.trip_infos)

            old_route_infos_len = self.route_infos_len
            self.route_info_db.route_infos = (
                self.route_info_db.route_infos | rti_db_task.result().route_infos
            )
            self.route_infos_len = len(self.route_info_db.route_infos)

            _LOGGER.info(
                f"GTFS Static Coordinator Merge New Data Update Complete {self.gtfs_static_zip}"
            )
            _LOGGER.info(
                f"GTFS Static Coordinator Services: {old_services_len} -> {self.services_len}"
            )
            _LOGGER.info(
                f"GTFS Static Coordinator Station Stop Infos: {old_stop_infos_len} -> {self.stop_infos_len}"
            )
            _LOGGER.info(
                f"GTFS Static Coordinator Trip Infos: {old_trip_infos_len} -> {self.trip_infos_len}"
            )
            _LOGGER.info(
                f"GTFS Static Coordinator Route Infos: {old_route_infos_len} -> {self.route_infos_len}"
            )
        self.last_static_update = datetime.now()


class GtfsStaticCoordinator(DataUpdateCoordinator):
    """GTFS Static Update Coordinator. Polls Static Data Endpoints for new data on a slower basis."""

    def __init__(
        self,
        hass: HomeAssistant,
    ) -> None:
        """Initialize the GTFS Update Coordinator to notify all entities upon poll."""
        super().__init__(
            hass, _LOGGER, name="GTFS Static", update_interval=timedelta(days=1)
        )
        # Save the resource path to reload periodically
        _LOGGER.info("Setup GTFS Static Update Coordinator")

    async def _async_update_data(self):
        """Fetch data from API endpoint."""
