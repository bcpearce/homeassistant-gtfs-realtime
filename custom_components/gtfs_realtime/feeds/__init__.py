"""Module for bootstrapping certain transit systems."""
