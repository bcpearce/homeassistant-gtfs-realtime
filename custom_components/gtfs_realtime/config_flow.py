import json
import logging
from typing import Any

import aiohttp
from gtfs_station_stop.route_info import RouteInfoDatabase
from gtfs_station_stop.static_database import async_factory
from gtfs_station_stop.station_stop_info import LocationType, StationStopInfoDatabase
from homeassistant import config_entries
import homeassistant.helpers.config_validation as cv
from homeassistant.helpers.selector import (
    SelectOptionDict,
    SelectSelector,
    SelectSelectorConfig,
    SelectSelectorMode,
    TextSelector,
    TextSelectorConfig,
    TextSelectorType,
)
import voluptuous as vol

from .const import (
    CONF_API_KEY,
    CONF_ARRIVAL_LIMIT,
    CONF_GTFS_PROVIDER,
    CONF_GTFS_STATIC_DATA,
    CONF_ROUTE_ICONS,
    CONF_ROUTE_IDS,
    CONF_STOP_IDS,
    CONF_URL_ENDPOINTS,
    DOMAIN,
)

DOMAIN_SCHEMA = vol.Schema(
    {
        vol.Optional(CONF_API_KEY): cv.string,
        vol.Required(CONF_URL_ENDPOINTS): vol.All([cv.url]),
        vol.Optional(CONF_GTFS_STATIC_DATA): vol.All([cv.url]),
        vol.Optional(CONF_ROUTE_ICONS): cv.path,
    }
)


_LOGGER = logging.getLogger(__name__)


class GtfsRealtimeConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
    """Config flow for GTFS Realtime"""

    VERSION = 1
    FEEDS_URL = "https://gist.githubusercontent.com/bcpearce/cc60c18f4022c4a11c460c5ccd2ec158/raw/feeds.json"
    feeds = {}

    def __init__(self) -> None:
        """Initialize config flow."""
        self.device_config: dict[str, Any] = {}

    async def get_feeds():
        async with aiohttp.ClientSession() as session:
            async with session.get(GtfsRealtimeConfigFlow.FEEDS_URL) as response:
                if response.status >= 200 and response.status < 400:
                    GtfsRealtimeConfigFlow.feeds = json.loads(await response.text())

    async def async_step_user(self, user_input=None):
        errors = {}
        if user_input is not None:
            if CONF_GTFS_PROVIDER in user_input:
                return await self.async_step_choose_static_and_realtime_feeds(
                    user_input
                )

        # Update the feeds, typically this will be from an externally hosted
        # file so it may be kept up to date without requiring updates to this repository.
        # It can also be monkey patched to support testing.
        try:
            await GtfsRealtimeConfigFlow.get_feeds()
        except Exception as e:
            # do not allow errors to propagate, this is for convenience
            _LOGGER.error("Failed to get preconfigured feeds")
            errors["base"] = f"Failed to get preconfigured feeds: {e}"

        options = {"_": "Other - Enter Manually"}
        for k, v in GtfsRealtimeConfigFlow.feeds.items():
            options[k] = v["name"]

        data_schema = vol.Schema(
            {
                vol.Required(CONF_GTFS_PROVIDER): SelectSelector(
                    SelectSelectorConfig(
                        mode=SelectSelectorMode.DROPDOWN,
                        options=[
                            SelectOptionDict(value=k, label=v)
                            for k, v in options.items()
                        ],
                    )
                ),
            }
        )
        return self.async_show_form(
            step_id="user", data_schema=data_schema, errors=errors
        )

    async def async_step_choose_static_and_realtime_feeds(
        self, user_input: dict[str, str], errors: dict[str, str] = {}
    ):
        if (
            CONF_GTFS_STATIC_DATA in user_input
            and CONF_URL_ENDPOINTS in user_input
            and not errors
        ):
            self.device_config = self.device_config | user_input
            return await self.async_step_choose_informed_entities(user_input)
        gtfs_provider = user_input.get(CONF_GTFS_PROVIDER)
        realtime_feeds = [""]
        static_feeds = [""]
        route_icons = ""
        self.device_config[CONF_GTFS_PROVIDER] = "Manual"
        feed_data = GtfsRealtimeConfigFlow.feeds.get(gtfs_provider)
        if feed_data is not None:
            realtime_feeds = list(feed_data["realtime_feeds"].values())
            static_feeds = list(feed_data["static_feeds"].values())
            route_icons = feed_data.get("route_icons", route_icons)
            self.device_config[CONF_GTFS_PROVIDER] = feed_data["name"]

        data_schema = vol.Schema(
            {
                vol.Optional(CONF_API_KEY, default=""): cv.string,
                vol.Optional(
                    CONF_URL_ENDPOINTS,
                    default=realtime_feeds,
                    description={"suggested_value": ["https://"]}
                    if realtime_feeds == [""]
                    else {},
                ): TextSelector(
                    TextSelectorConfig(
                        multiline=False,
                        type=TextSelectorType.URL,
                        multiple=True,
                    )
                ),
                vol.Optional(
                    CONF_GTFS_STATIC_DATA,
                    default=static_feeds,
                    description={"suggested_value": ["https://"]}
                    if static_feeds == [""]
                    else {},
                ): TextSelector(TextSelectorConfig(multiline=False, multiple=True)),
                vol.Optional(
                    CONF_ROUTE_ICONS,
                    default=route_icons,
                ): cv.string,
            }
        )
        return self.async_show_form(
            step_id="choose_static_and_realtime_feeds",
            data_schema=data_schema,
            errors=errors,
        )

    async def async_step_choose_informed_entities(self, user_input):
        if CONF_ROUTE_IDS in user_input or CONF_STOP_IDS in user_input:
            self.device_config = self.device_config | user_input
            return self.async_create_entry(
                title=user_input[CONF_GTFS_PROVIDER],
                data=self.device_config,
            )
        errors = {}
        headers = {}
        if user_input.get(CONF_API_KEY):
            headers["api_key"] = user_input[CONF_API_KEY]
        try:
            ssi_db = await async_factory(
                StationStopInfoDatabase,
                *user_input[CONF_GTFS_STATIC_DATA],
                headers=headers,
            )
            rt_db = await async_factory(
                RouteInfoDatabase, *user_input[CONF_GTFS_STATIC_DATA], headers=headers
            )
        except Exception as e:
            errors["base"] = str(e)
            return await self.async_step_choose_static_and_realtime_feeds(
                {CONF_GTFS_PROVIDER: self.device_config.get(CONF_GTFS_PROVIDER)}, errors
            )
        stops = [
            SelectOptionDict(
                value=k,
                label=f"{v.name} {f' - {v.desc}' if v.desc is not None else ''} ({v.id})",
            )
            for k, v in ssi_db.station_stop_infos.items()
            if v.location_type == LocationType.STOP
        ]
        data_schema = vol.Schema(
            {
                vol.Required(
                    CONF_GTFS_PROVIDER,
                    default=self.device_config.get(
                        CONF_GTFS_PROVIDER, "Generic GTFS Provider"
                    ),
                ): cv.string,
                vol.Optional(CONF_ROUTE_IDS): SelectSelector(
                    SelectSelectorConfig(
                        options=[
                            SelectOptionDict(
                                value=k,
                                label=f"{k}: {rt_db.route_infos[k].long_name or rt_db.route_infos[k].short_name}",
                            )
                            for k in rt_db.route_infos.keys()
                        ],
                        mode=SelectSelectorMode.DROPDOWN,
                        multiple=True,
                    )
                ),
                vol.Optional(CONF_STOP_IDS): SelectSelector(
                    SelectSelectorConfig(
                        options=stops,
                        mode=SelectSelectorMode.DROPDOWN,
                        multiple=True,
                    )
                ),
                vol.Required(CONF_ARRIVAL_LIMIT, default=4): int,
            }
        )

        return self.async_show_form(
            step_id="choose_informed_entities", data_schema=data_schema, errors=errors
        )
