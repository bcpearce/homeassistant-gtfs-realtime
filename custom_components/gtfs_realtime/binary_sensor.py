"""Platform for binary sensor integration."""

from __future__ import annotations

from gtfs_station_stop.route_status import RouteStatus
from gtfs_station_stop.station_stop import StationStop
from gtfs_station_stop.station_stop_info import StationStopInfo
from homeassistant.components.binary_sensor import (
    PLATFORM_SCHEMA as BINARY_SENSOR_PLATFORM_SCHEMA,
    BinarySensorDeviceClass,
    BinarySensorEntity,
)
from homeassistant.core import HomeAssistant, callback
import homeassistant.helpers.config_validation as cv
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType
from homeassistant.helpers.update_coordinator import CoordinatorEntity
import voluptuous as vol

from .const import (
    CONF_ROUTE_IDS,
    COORDINATOR_REALTIME,
    DESCRIPTION_PRETTY,
    DOMAIN,
    HEADER_PRETTY,
    ROUTE_ID,
    STOP_ID,
)
from .coordinator import GtfsRealtimeCoordinator

PLATFORM_SCHEMA = BINARY_SENSOR_PLATFORM_SCHEMA.extend(
    {
        vol.Required(
            vol.Any(ROUTE_ID, STOP_ID),
            msg=f"Must specify at least one of ['{ROUTE_ID}', 'f{STOP_ID}']",
        ): cv.string
    }
)


async def async_setup_entry(
    hass: HomeAssistant,
    config: ConfigType,
    add_entities: AddEntitiesCallback,
    discovery_info: DiscoveryInfoType | None = None,
) -> None:
    """Set up the sensor platform."""
    coordinator: GtfsRealtimeCoordinator = hass.data[DOMAIN][COORDINATOR_REALTIME]
    if discovery_info is None:
        if CONF_ROUTE_IDS in config.data:
            add_entities(
                [
                    AlertSensor(
                        coordinator,
                        RouteStatus(route_id, coordinator.hub),
                        hass.config.language,
                        None,
                    )
                    for route_id in config.data[CONF_ROUTE_IDS]
                ]
            )


class AlertSensor(BinarySensorEntity, CoordinatorEntity):
    """Representation of a Station GTFS Realtime Alert Sensor."""

    CLEAN_ALERT_DATA = {HEADER_PRETTY: "", DESCRIPTION_PRETTY: ""}

    _attr_device_class = BinarySensorDeviceClass.PROBLEM

    def __init__(
        self,
        coordinator: GtfsRealtimeCoordinator,
        informed_entity: StationStop | RouteStatus,
        language: str = "",
        station_stop_info: StationStopInfo | None = None,
    ) -> None:
        """Initialize the sensor."""
        super().__init__(coordinator)
        self.informed_entity = informed_entity
        self.language = language
        self._name: str = f"{station_stop_info.name if station_stop_info is not None else informed_entity.id} Service Alerts"
        self._attr_is_on = False
        self._alert_detail: dict[str, str] = AlertSensor.CLEAN_ALERT_DATA
        self._attr_unique_id = f"alert_{informed_entity.id}"

    @property
    def name(self) -> str | None:
        """Name of the Sensor."""
        return self._name

    @property
    def extra_state_attributes(self) -> dict[str, str]:
        """Explanation of Alerts for a given Stop ID."""
        return self._alert_detail

    def update(self) -> None:
        alerts = self.informed_entity.alerts
        self._alert_detail = {}
        if len(alerts) == 0:
            self._attr_is_on = False
        elif len(alerts) > 0:
            self._attr_is_on = True
            for i, alert in enumerate(alerts):
                self._alert_detail[f"{HEADER_PRETTY}{f" {i + 1}" if i > 0 else ""}"] = (
                    alert.header_text.get(self.language, "")
                )
                self._alert_detail[
                    f"{DESCRIPTION_PRETTY}{f" {i + 1}" if i > 0 else ""}"
                ] = alert.description_text.get(self.language, "")
        self.async_write_ha_state()

    @callback
    def _handle_coordinator_update(self) -> None:
        self.update()
