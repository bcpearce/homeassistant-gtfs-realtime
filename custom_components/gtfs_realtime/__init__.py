"""The GTFS Realtime integration."""
# GTFS Station Stop Feed Subject serves as the data hub for the integration

from datetime import timedelta
from typing import Any

from gtfs_station_stop.feed_subject import FeedSubject
from homeassistant.config_entries import ConfigEntry
from homeassistant.const import Platform
from homeassistant.core import HomeAssistant
import voluptuous as vol

from custom_components.gtfs_realtime.config_flow import DOMAIN_SCHEMA

from .const import (
    CAL_DB,
    CONF_API_KEY,
    CONF_GTFS_STATIC_DATA,
    CONF_ROUTE_ICONS,
    CONF_URL_ENDPOINTS,
    COORDINATOR_REALTIME,
    DOMAIN,
    RTI_DB,
    SSI_DB,
    TI_DB,
)
from .coordinator import GtfsRealtimeCoordinator

PLATFORMS = [Platform.BINARY_SENSOR, Platform.SENSOR]

CONFIG_SCHEMA = vol.Schema(
    {DOMAIN: DOMAIN_SCHEMA},
    extra=vol.ALLOW_EXTRA,
)


async def _async_create_gtfs_update_hub(hass: HomeAssistant, config: dict[str, Any]):
    hub = FeedSubject(
        config[CONF_URL_ENDPOINTS], headers={"api_key": config[CONF_API_KEY]}
    )
    route_icons = config.get(CONF_ROUTE_ICONS)  # optional
    # Attempt to perform an update to verify configuration
    await hub.async_update()
    coordinator_realtime = GtfsRealtimeCoordinator(
        hass, hub, config[CONF_GTFS_STATIC_DATA], static_timedelta=timedelta(hours=24)
    )
    # Update the static data for the coordinator before the first update
    await coordinator_realtime.async_update_static_data()
    hass.data[DOMAIN] = {
        COORDINATOR_REALTIME: coordinator_realtime,
        CAL_DB: coordinator_realtime.calendar,
        SSI_DB: coordinator_realtime.station_stop_info_db,
        TI_DB: coordinator_realtime.trip_info_db,
        RTI_DB: coordinator_realtime.route_info_db,
        CONF_ROUTE_ICONS: route_icons,
    }
    if CONF_ROUTE_ICONS in config:
        hass.data[DOMAIN][CONF_ROUTE_ICONS] = config[CONF_ROUTE_ICONS]
    return True


async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
    """Set up GTFS Realtime Feed Subject for use by all sensors."""
    await _async_create_gtfs_update_hub(hass, entry.data)
    await hass.config_entries.async_forward_entry_setups(entry, PLATFORMS)
    return True
