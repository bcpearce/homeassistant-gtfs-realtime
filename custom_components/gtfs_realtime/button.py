"""Button support for GTFS Schedule Update."""

from __future__ import annotations

from collections.abc import Callable, Coroutine
from dataclasses import dataclass
from functools import cached_property
import os

from homeassistant.components.button import (
    ButtonDeviceClass,
    ButtonEntity,
    ButtonEntityDescription,
)
from homeassistant.const import EntityCategory
from homeassistant.core import HomeAssistant
from homeassistant.helpers.device_registry import DeviceInfo
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.update_coordinator import CoordinatorEntity

from custom_components.gtfs_realtime import GtfsRealtimeConfigEntry
from custom_components.gtfs_realtime.const import DOMAIN

from .coordinator import GtfsRealtimeCoordinator


@dataclass(frozen=True, kw_only=True)
class GtfsRealtimeButtonDescription(ButtonEntityDescription):
    """GTFS Realtime Button description."""

    press_action: Callable[[GtfsRealtimeCoordinator, os.PathLike], Coroutine]


async def handle_refresh_button_press(
    coordinator: GtfsRealtimeCoordinator, gtfs_static_source: os.PathLike
):
    """Called on pressing refresh feed entity."""
    coordinator.static_update_targets.add(gtfs_static_source)
    await coordinator.async_update_static_data()


BUTTONS: dict[str, GtfsRealtimeButtonDescription] = {
    "refresh": GtfsRealtimeButtonDescription(
        key="refresh",
        translation_key="refresh",
        device_class=ButtonDeviceClass.UPDATE,
        entity_category=EntityCategory.CONFIG,
        press_action=handle_refresh_button_press,
    ),
    "clear": GtfsRealtimeButtonDescription(
        key="clear",
        translation_key="clear",
        device_class=ButtonDeviceClass.UPDATE,
        entity_category=EntityCategory.CONFIG,
        press_action=lambda coordinator: coordinator.async_update_static_data(
            clear_old_data=True
        ),
    ),
}


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: GtfsRealtimeConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    coordinator = config_entry.runtime_data
    async_add_entities(
        [
            ScheduleUpdateButton(coordinator, BUTTONS["refresh"], gtfs_static_source)
            for gtfs_static_source in coordinator.gtfs_static_zip
        ]
        + [ScheduleClearButton(coordinator, BUTTONS["clear"])]
    )


class ScheduleButtonBase(ButtonEntity, CoordinatorEntity):
    entity_description: GtfsRealtimeButtonDescription

    def __init__(
        self,
        coordinator: GtfsRealtimeCoordinator,
        description: GtfsRealtimeButtonDescription,
    ):
        super().__init__(coordinator)
        self.entity_description = description

    @cached_property
    def device_info(self) -> DeviceInfo:
        """Return the device info."""
        return DeviceInfo(
            identifiers={(DOMAIN, ",".join(self.coordinator.gtfs_static_zip))},
            name="GTFS Schedule",
            manufacturer=self.coordinator.gtfs_provider,
        )


class ScheduleUpdateButton(ScheduleButtonBase):
    def __init__(
        self,
        coordinator: GtfsRealtimeCoordinator,
        description: GtfsRealtimeButtonDescription,
        gtfs_static_source: os.PathLike | None = None,
    ):
        super().__init__(coordinator, description)
        self.gtfs_static_source = gtfs_static_source
        self._attr_unique_id = f"refresh_gtfs_schedule_button-{gtfs_static_source}"
        self._attr_translation_key = description.translation_key
        # Figure out how to use translations for this
        self._attr_name = f"Refresh Schedule Feed: {self.gtfs_static_source}"

    @cached_property
    def extra_state_attributes(self) -> dict[str, str]:
        return {"target_url": str(self.gtfs_static_source)}

    async def async_press(self) -> None:
        """Trigger the button action."""
        await self.entity_description.press_action(
            self.coordinator, self.gtfs_static_source
        )


class ScheduleClearButton(ScheduleButtonBase):
    entity_description: GtfsRealtimeButtonDescription

    def __init__(
        self,
        coordinator: GtfsRealtimeCoordinator,
        description: GtfsRealtimeButtonDescription,
    ):
        super().__init__(coordinator, description)
        self._attr_unique_id = (
            f"clear_gtfs_schedule-{"-".join(coordinator.gtfs_static_zip)}"
        )
        self._attr_translation_key = description.translation_key
        self._attr_name = "Clear GTFS Schedule"

    @cached_property
    def extra_state_attributes(self) -> dict[str, str]:
        return {
            f"target_url {i+1}": url
            for i, url in enumerate(self.coordinator.gtfs_static_zip)
        }

    async def async_press(self) -> None:
        """Trigger the button action."""
        await self.entity_description.press_action(self.coordinator)
