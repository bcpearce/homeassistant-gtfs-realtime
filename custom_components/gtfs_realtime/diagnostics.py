from dataclasses import asdict
from typing import Any

from homeassistant.core import HomeAssistant

from custom_components.gtfs_realtime import GtfsRealtimeConfigEntry


async def async_get_config_entry_diagnostics(
    hass: HomeAssistant, entry: GtfsRealtimeConfigEntry
) -> dict[str, Any]:
    return {
        "schedule": asdict(entry.runtime_data.data.schedule),
        "last_static_update": entry.runtime_data.last_static_update,
        "static_update_frequency": entry.runtime_data.static_timedelta,
    }
