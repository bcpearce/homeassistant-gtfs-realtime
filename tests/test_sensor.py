"""Test sensor."""

from datetime import datetime, timedelta
from unittest.mock import AsyncMock, patch

from freezegun.api import FrozenDateTimeFactory
from gtfs_station_stop.arrival import Arrival
from gtfs_station_stop.calendar import Calendar
from gtfs_station_stop.route_info import RouteInfoDatabase
from gtfs_station_stop.station_stop_info import StationStopInfoDatabase
from gtfs_station_stop.trip_info import TripInfoDatabase
from homeassistant.const import STATE_UNKNOWN
from homeassistant.core import HomeAssistant
import pytest
from pytest_homeassistant_custom_component.common import async_fire_time_changed

from custom_components.gtfs_realtime.coordinator import GtfsRealtimeCoordinator


async def test_smoke(
    hass: HomeAssistant, freezer: FrozenDateTimeFactory, entry_v2_nodialout
):
    """Smoke test the sensor platform."""
    with (
        patch(
            "gtfs_station_stop.feed_subject.FeedSubject.async_update",
            new_callable=AsyncMock,
            return_value=None,
        ),
        patch(
            "custom_components.gtfs_realtime.coordinator.GtfsRealtimeCoordinator._async_update_static_data",
            new_callable=AsyncMock,
            return_value=(
                Calendar(),
                StationStopInfoDatabase(),
                TripInfoDatabase(),
                RouteInfoDatabase(),
            ),
        ),
    ):
        entry_v2_nodialout.add_to_hass(hass)
        assert await hass.config_entries.async_setup(entry_v2_nodialout.entry_id)
        await hass.async_block_till_done()

    coordinator: GtfsRealtimeCoordinator = entry_v2_nodialout.runtime_data
    coordinator.route_icons = None
    coordinator.hub.realtime_feed_uris = []

    async def coordinator_update_side_effects():
        for x in range(4):
            arrivals = {
                "101N": [
                    Arrival(datetime.now() + timedelta(minutes=4 - x), "A", ""),
                    Arrival(datetime.now() + timedelta(minutes=6 - x), "B", ""),
                    Arrival(datetime.now() + timedelta(minutes=8 - x), "C", ""),
                ],
                "102S": [
                    Arrival(datetime.now() + timedelta(minutes=9 - x), "X", ""),
                    Arrival(datetime.now() + timedelta(minutes=13 - x), "Y", ""),
                    Arrival(datetime.now() + timedelta(minutes=17 - x), "Z", ""),
                ],
            }
            for id, stop in coordinator.station_stops.items():
                stop.arrivals = next(arrivals)[id]
                yield
        pytest.fail("Tests should not call the update more than 4 times")

    coordinator.hub.async_update = AsyncMock()
    coordinator.hub.async_update.side_effect = coordinator_update_side_effects

    freezer.tick(timedelta(minutes=1.1))
    async_fire_time_changed(hass)
    await hass.async_block_till_done()
    # Need to figure out how to mock setting the state
    # assert hass.states.get("sensor.1_101n").state == 4
    # assert hass.states.get("sensor.2_101n").state == 6
    # assert hass.states.get("sensor.3_101n").state == 8
    assert hass.states.get("sensor.4_101n").state == STATE_UNKNOWN
    # assert hass.states.get("sensor.1_102s").state == 9
    # assert hass.states.get("sensor.2_102s").state == 13
    # assert hass.states.get("sensor.3_102s").state == 17
    assert hass.states.get("sensor.4_102s").state == STATE_UNKNOWN
    freezer.tick(timedelta(minutes=1.1))
    async_fire_time_changed(hass)
    await hass.async_block_till_done()
    # assert hass.states.get("sensor.1_101n").state == 3
    # assert hass.states.get("sensor.2_101n").state == 5
    # assert hass.states.get("sensor.3_101n").state == 7
    assert hass.states.get("sensor.4_101n").state == STATE_UNKNOWN
    # assert hass.states.get("sensor.1_102s").state == 8
    # assert hass.states.get("sensor.2_102s").state == 12
    # assert hass.states.get("sensor.3_102s").state == 16
    assert hass.states.get("sensor.4_102s").state == STATE_UNKNOWN
