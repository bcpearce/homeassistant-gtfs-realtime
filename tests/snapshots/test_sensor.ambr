# serializer version: 1
# name: test_update[1_101n-after-1-minute]
  '3.0'
# ---
# name: test_update[1_102s-after-1-minute]
  '8.0'
# ---
# name: test_update[2_101n-after-1-minute]
  '5.0'
# ---
# name: test_update[2_102s-after-1-minute]
  '12.0'
# ---
# name: test_update[3_101n-after-1-minute]
  '7.0'
# ---
# name: test_update[3_102s-after-1-minute]
  '16.0'
# ---
# name: test_update[4_101n-after-1-minute]
  'unknown'
# ---
# name: test_update[4_102s-after-1-minute]
  'unknown'
# ---
